File: 332. Reconstruct Itinerary.py
class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        self.n = len(tickets)
        from_to = {}
        visited = {}
        for ticket in tickets:
            if ticket[0] not in from_to:
                from_to[ticket[0]] = []
                visited[ticket[0]] = []
            from_to[ticket[0]] += [ticket[1]]
            visited[ticket[0]] += [False]
        for key in from_to.keys():
            from_to[key] = sorted(from_to[key])
        self.res = []
        self.DFS(['JFK'], 'JFK', from_to, visited)
        return self.res
        
    def DFS(self, path, start, from_to, visited):
        if len(path) == self.n + 1:
            self.res = path[:]
            return
        if not self.res and start in from_to:
            for e in xrange(len(from_to[start])):
                if not visited[start][e]:
                    visited[start][e] = True
                    self.DFS(path + [from_to[start][e]], from_to[start][e], from_to, visited)
                    visited[start][e] = False

File: 332. Reconstruct Itinerary.js
/**
 * @param {string[][]} tickets
 * @return {string[]}
 */
var findItinerary = function(tickets) {
    let from_to = {}, visited = {};
    for (let i = 0; i < tickets.length; i++)
    {
        let start = tickets[i][0], end = tickets[i][1];
        if (!(start in from_to))
        {
            from_to[start] = [];
            visited[start] = [];
        }
        from_to[start].push(end);
        visited[start].push(false);
    }
    for (let start in from_to)
        from_to[start] = from_to[start].sort();
    
    let res = [];
    var DFS = function(path, start, visited)
    {
        if (path.length == tickets.length + 1)
        {
            for (let i = 0; i < path.length; i++)
                res.push(path[i]);
            return;
        }
        if (res.length === 0 && start in from_to)
        {
            for (let e = 0; e < from_to[start].length; e++)
            {
                if (!visited[start][e])
                {
                    visited[start][e] = true;
                    path.push(from_to[start][e]);
                    DFS(path, from_to[start][e], visited);
                    path.pop();
                    visited[start][e] = false;
                }
            }
        }
    };
    DFS(['JFK'], 'JFK', visited);
    return res;
};
